gsy-matching-engine/src/connectors/substrate_connector.rs

```diff
@@ -0,0 +1,193 @@
+use crate::algorithms::PayAsBid;
+use crate::primitives::web3::{
+    Bid, BidOfferMatch, MatchingData, Offer, Order, OrderSchema, OrderStatus,
+};
+use crate::primitives::web3_extension::BidOfferMatch as BidOfferMatchExtension;
+use anyhow::{Error, Result};
+use async_recursion::async_recursion;
+use codec::{Decode, Encode};
+use sp_keyring::AccountKeyring;
+use std::sync::{Arc, Mutex};
+use std::{thread, time};
+use subxt::sp_runtime::AccountId32;
+use subxt::{
+    rpc::Subscription,
+    sp_runtime::{generic::Header, traits::BlakeTwo256},
+    ClientBuilder, DefaultConfig, PairSigner, PolkadotExtrinsicParams, SubstrateExtrinsicParams,
+};
+use tracing::{error, info};
+
+#[subxt::subxt(runtime_metadata_path = "metadata.scale")]
+pub mod gsy_node {}
+
+pub const DEFAULT_MARKET_ID: u8 = 1;
+
+use crate::connectors::substrate_connector::gsy_node::runtime_types::gsy_primitives::trades::BidOfferMatch as OtherBidOfferMatch;
+
+#[async_recursion]
+pub async fn substrate_subscribe(orderbook_url: String, node_url: String) -> Result<(), Error> {
+    info!("Connecting to {}", node_url);
+
+    let api = ClientBuilder::new()
+        .set_url(node_url.clone())
+        .build()
+        .await?
+        .to_runtime_api::<gsy_node::RuntimeApi<DefaultConfig, SubstrateExtrinsicParams<DefaultConfig>>>();
+
+    let mut gsy_blocks_events: Subscription<Header<u32, BlakeTwo256>> =
+        api.client.rpc().subscribe_finalized_blocks().await?;
+
+    let orderbook_url = Arc::new(Mutex::new(orderbook_url));
+    let node_url = Arc::new(Mutex::new(node_url.clone()));
+
+    while let Some(Ok(block)) = gsy_blocks_events.next().await {
+        info!("Block {:?} finalized: {:?}", block.number, block.hash());
+
+        let matches = Arc::new(Mutex::new(Vec::new()));
+
+        if (block.number as u64) % 4 == 0 {
```

Comment:
Can you please explain why you do a mod 4 here? Thanks!

Reply:
This is trying to configure the matching frequency actually. Essentially it says that it will perform the matching operation every 4 blocks. I will rename it in order to be more easily readable